### PR TITLE
PB-1051: QR Codes - ability to customise barcode colours added [PREVIEW PR, DO NOT MERGE]

### PR DESCRIPTION
**Pull request by peteward**

### CODE REVIEW

File: packages/common/src/helper.ts [hunk 1]
```diff
@@ -193,11 +193,12 @@ export const validateBarcodeInput = (type: BarCodeType, input: string) => {
     let res = input.match(regexp);
     if (
       res != null &&
+      input.length <= 52 && // 52 is the max length of a GS1 DataMatrix barcode before bwip-js throaws an error
       res[1] === '01' &&
       (res[2].length === 14 || res[2].length === 8 || res[2].length === 12 || res[2].length === 13)
     ) {
       let gtin = res[2];
-      ret = validateCheckDigit(gtin, gtin.length);
+        ret = validateCheckDigit(gtin, gtin.length);
```
Comment (peteward) [inline packages/common/src/helper.ts]: revert

File: packages/generator/src/helper.ts [hunk 2]
```diff
@@ -55,17 +55,20 @@ export const createBarCode = async (arg: {
...
+  textcolor?: string;
 }): Promise<Buffer> => {
-  const { type, input, width, height, backgroundColor } = arg;
+  const { type, input, width, height, backgroundcolor, barcolor, textcolor } = arg;
+  
   const bcid = barCodeType2Bcid(type);
   const includetext = true;
   const scale = 5;
-  const bwipjsArg: ToBufferOptions = { bcid, text: input, width, height, scale, includetext };
+  const bwipjsArg: ToBufferOptions = { bcid, text: input, width, height, scale, includetext, };
```
Comment (peteward) [inline packages/generator/src/helper.ts]: revert

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+const inputStyle = {
+  width: '90%',
+  color: '#333',
+  background: 'none',
+  borderRadius: 2,
+  border: '1px solid #767676',
+};
+
+const ColorInputSet = (props: {
```
Comment (peteward) [inline packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: I see that there's already a ColorInputSet, which is at least very similar to what you're doing here, in the `TextPropEditor`. 

I think we should consolidate this code and make it shared.

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          }}
+          onClear={() => {
+            changeSchemas([{ key: 'backgroundcolor', value: '', schemaId: activeSchema.id }])
+          }}
+        />
+
+        {barcodeHasText && (
+          <ColorInputSet
+            label={i18n('textColor')}
+            value={activeSchema.textcolor ?? '#000000'}
```
Comment (peteward) [inline packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: use `DEFAULT_FONT_COLOR` as the fallback here?

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+          marginBottom: '1rem',
+          display: 'flex',
+          alignItems: 'center',
+          justifyContent: 'space-between',
+          flexWrap: 'wrap',
+        }}
+      >
+        <ColorInputSet
+          label={i18n('barColor')}
+          value={activeSchema.barcolor ?? '#000000'}
```
Comment (peteward) [inline packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: We should probably add a constant for this default

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          onClear={() =>
+            changeSchemas([
+              { key: 'barcolor', value: DEFAULT_FONT_COLOR, schemaId: activeSchema.id },
+            ])
+          }
+        />
+
+        <ColorInputSet
+          label={i18n('bgColor')}
+          value={activeSchema.backgroundcolor ?? '#ffffff'}
```
Comment (peteward) [inline packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: same here

File: packages/ui/src/components/Schemas/BarcodeSchema.tsx [hunk 7]
```diff
@@ -1,39 +1,20 @@
...
-      width: schema.width * ZOOM,
-      height: schema.height * ZOOM,
-      position: 'absolute',
-      borderRadius: 0,
-      opacity: 0.5,
-    }}
-    src={barcodeExampleImageObj[schema.type]}
-  />
+const SampleBarcode = ({ schema, imageSrc }: { schema: BarcodeSchema, imageSrc: string }) => (
+      <img
```
Comment (peteward) [inline packages/ui/src/components/Schemas/BarcodeSchema.tsx]: the indents don't look right here, should probably be 2 indents for this line then reduce the others also?

**Review by peteward — COMMENTED** [review]
Excellent work!

A couple of tiny points then a little refactor of the helper.
I might pick this up if I get a bit of time (no need for you on your holidays!) then you can open up the resulting PR upstream.

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx (hunk 3 above)
Comment (peteward) [reply packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: actually, whilst this should be done, it opens some design decisions that can happen separately. 
We can comment about this on the upstream PR 👍

**Comment by peteward** [comment]
@steffancarrington I have applied the review fixes. 
I also spotted a bug in rendering an empty QR code on the FE and fixed it.
Check the details on this commit: https://github.com/pennyblack-io/pdfme/pull/4/commits/a4311442340125132462d0a2b9d20aa3f9697486

I think this is good to go for an upstream PR 👍

File: packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx (hunk 3 above)
Comment (steffancarrington) [reply packages/ui/src/components/Designer/Sidebar/DetailView/BarCodePropEditor.tsx]: Ah yes, this did cross my mind, the src directory could do with a bit of a clean-up and start making prop editor components shared across the UI. 👍

**Pull request closed without merging**